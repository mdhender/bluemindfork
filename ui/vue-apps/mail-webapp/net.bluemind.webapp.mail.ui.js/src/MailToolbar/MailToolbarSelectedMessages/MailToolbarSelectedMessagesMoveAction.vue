<template>
    <bm-dropdown
        ref="move-dropdown"
        v-bm-tooltip.bottom
        no-caret
        class="h-100 move-message"
        variant="simple-dark"
        :title="$t('mail.toolbar.move.tooltip')"
        :aria-label="$t('mail.actions.move.aria')"
        @shown="openMoveAutocomplete"
        @hide="resetPattern"
    >
        <global-events @keydown.tab.capture="forceCloseMoveAutocomplete" />
        <template slot="button-content">
            <bm-icon icon="folder" size="2x" />
            <span class="d-none d-lg-block"> {{ $t("mail.actions.move") }}</span>
        </template>
        <bm-dropdown-autocomplete
            ref="moveAutocomplete"
            v-slot="{ item }"
            v-model.trim="pattern"
            :items="matchingFolders"
            search-icon
            :max-results="maxFolders"
            has-divider-under-input
            @keydown.esc.native="resetPattern"
        >
            <bm-dropdown-item-button
                v-bm-tooltip.left
                class="text-nowrap text-truncate w-100"
                :title="$tc('mail.actions.move.item', 1, { path: item.path })"
                @click="selectFolder(item)"
            >
                <template #icon>
                    <mail-folder-icon no-text :shared="isFolderOfMailshare(item)" :folder="item" />
                </template>
                {{ item.path }}
            </bm-dropdown-item-button>
        </bm-dropdown-autocomplete>
        <bm-dropdown-divider />
        <bm-dropdown-form
            v-if="pattern === ''"
            v-bm-tooltip.left
            class="new-folder position-relative"
            :aria-label="$t('mail.actions.create.folder')"
            :title="$t('mail.actions.create.folder')"
        >
            <mail-folder-input
                class="pl-2 pr-1"
                :submit-on-focusout="false"
                @submit="newFolderName => selectFolder({ name: newFolderName, path: newFolderName })"
                @keydown.left.native.stop
                @keydown.right.native.stop
                @keydown.esc.native.stop
            />
        </bm-dropdown-form>
        <bm-dropdown-item-button
            v-else-if="displayCreateFolderBtnFromPattern"
            v-bm-tooltip.left
            :aria-label="$tc('mail.actions.move.item', 1, { path: pattern })"
            :title="$tc('mail.actions.move.item', 1, { path: pattern })"
            icon="plus"
            @click="selectFolder({ name: pattern, path: pattern })"
        >
            {{ $t("mail.folder.new.from_pattern", [pattern]) }}
        </bm-dropdown-item-button>
    </bm-dropdown>
</template>

<script>
import {
    BmDropdown,
    BmDropdownAutocomplete,
    BmDropdownDivider,
    BmDropdownForm,
    BmDropdownItemButton,
    BmIcon,
    BmTooltip
} from "@bluemind/styleguide";
import { mapActions, mapGetters, mapState } from "vuex";
import GlobalEvents from "vue-global-events";
import MailFolderIcon from "../../MailFolderIcon";
import MailFolderInput from "../../MailFolderInput";
import ItemUri from "@bluemind/item-uri";

export default {
    name: "MailToolbarConsultMessageMoveAction",
    components: {
        BmDropdown,
        BmDropdownAutocomplete,
        BmDropdownDivider,
        BmDropdownForm,
        BmDropdownItemButton,
        BmIcon,
        GlobalEvents,
        MailFolderIcon,
        MailFolderInput
    },
    directives: { BmTooltip },
    data() {
        return {
            maxFolders: 10,
            pattern: ""
        };
    },
    computed: {
        ...mapState("mail-webapp", ["currentFolderKey"]),
        ...mapState("mail-webapp/currentMessage", { currentMessageKey: "key" }),
        ...mapGetters("mail-webapp", ["nextMessageKey", "my"]),
        ...mapState("mail", ["folders", "mailboxes"]),
        displayCreateFolderBtnFromPattern() {
            let pattern = this.pattern;
            if (pattern !== "") {
                pattern = pattern.replace(/\/+/, "/").replace(/^\/?(.*)\/?$/g, "$1");
                return (
                    pattern && !this.matchingFolders.some(match => match.path.toLowerCase() === pattern.toLowerCase())
                );
            }
            return false;
        },
        matchingFolders() {
            if (this.pattern !== "") {
                const filtered = Object.values(this.folders).filter(folder =>
                    folder.path.toLowerCase().includes(this.pattern.toLowerCase())
                );
                if (filtered) {
                    return filtered.slice(0, this.maxFolders);
                }
            }
            return [this.my.INBOX, this.my.TRASH].filter(
                folder => folder && ItemUri.encode(folder.key, folder.mailbox) !== this.currentFolderKey
            );
        }
    },
    methods: {
        ...mapActions("mail-webapp", ["move"]),
        selectFolder(item) {
            this.$router.navigate({ name: "v:mail:message", params: { message: this.nextMessageKey } });
            this.move({ messageKey: this.currentMessageKey, folder: item });
            this.$refs["move-dropdown"].hide(true);
        },
        openMoveAutocomplete() {
            this.$nextTick(() => this.$refs["moveAutocomplete"].focus());
        },
        forceCloseMoveAutocomplete() {
            this.$refs["move-dropdown"].hide(true);
            this.pattern = "";
        },
        resetPattern() {
            this.pattern = "";
        },
        isFolderOfMailshare(folder) {
            return this.mailboxes[folder.mailbox].type === "mailshares";
        }
    }
};
</script>

<style lang="scss">
@import "~@bluemind/styleguide/css/variables";
@import "~@bluemind/styleguide/css/mixins";

.move-message {
    input {
        border: none !important;
    }

    .dropdown-menu {
        min-width: 20vw;
        max-width: 40vw;
    }

    .bm-dropdown-autocomplete form {
        input {
            padding-left: $sp-3;
        }
        padding: 0 !important;
    }

    .dropdown-divider {
        margin: 0 !important;
    }

    .dropdown-item-content {
        @include text-overflow;
    }

    .dropdown-item,
    .b-dropdown-form {
        &:hover {
            background-color: $extra-light !important;
        }

        &:focus,
        &:focus:hover {
            background-color: $surface-active-bg !important;
        }
    }

    .b-dropdown-form {
        padding-top: $sp-1;
    }

    .new-folder .b-dropdown-form {
        padding-right: 0;
        padding-left: 0;
        padding-bottom: 0;
    }
}
</style>
