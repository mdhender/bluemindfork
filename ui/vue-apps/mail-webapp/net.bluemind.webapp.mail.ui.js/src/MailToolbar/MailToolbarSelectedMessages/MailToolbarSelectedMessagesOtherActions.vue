<template>
    <div class="d-inline-block h-100">
        <bm-dropdown
            v-bm-tooltip.bottom.ds500
            :no-caret="true"
            variant="link"
            :aria-label="$tc('mail.toolbar.more.aria')"
            :title="$tc('mail.toolbar.more.aria')"
            class="other_actions h-100"
            right
        >
            <template slot="button-content">
                <bm-icon icon="3dots" size="2x" />
                <span class="d-none d-lg-block">{{ $tc("mail.toolbar.more") }}</span>
            </template>
            <bm-dropdown-item class="shadow-sm" :data-shortcut="$t('mail.shortcuts.purge')" @click="deletionConfirmed">
                {{ $t("mail.actions.purge") }}
            </bm-dropdown-item>
        </bm-dropdown>
    </div>
</template>

<script>
import { BmDropdown, BmDropdownItem, BmIcon, BmTooltip } from "@bluemind/styleguide";
import { mapActions, mapGetters, mapState } from "vuex";

export default {
    name: "MailToolbarConsultMessageOtherActions",
    components: {
        BmDropdown,
        BmDropdownItem,
        BmIcon
    },
    directives: { BmTooltip },
    computed: {
        ...mapState("mail-webapp/currentMessage", { currentMessageKey: "key" }),
        ...mapGetters("mail-webapp", ["nextMessageKey"])
    },
    methods: {
        ...mapActions("mail-webapp", ["purge"]),
        deletionConfirmed() {
            this.$router.navigate({ name: "v:mail:message", params: { message: this.nextMessageKey } });
            this.purge(this.currentMessageKey);
        }
    }
};
</script>

<style lang="scss">
@import "~@bluemind/styleguide/css/_variables";

.other_actions .dropdown-menu {
    border: none !important;
    margin-top: $sp-1 !important;
    padding: 0 !important;
}
</style>
