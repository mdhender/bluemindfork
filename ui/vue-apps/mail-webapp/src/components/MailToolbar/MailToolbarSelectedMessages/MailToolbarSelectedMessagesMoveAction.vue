<template>
    <bm-dropdown
        ref="move-dropdown"
        no-caret
        class="h-100 move-message"
        variant="inline-light"
        toggle-class="btn-lg-simple-dark"
        :title="$t('mail.toolbar.move.tooltip')"
        :aria-label="$t('mail.actions.move.aria')"
        @shown="openMoveAutocomplete"
        @hide="resetPattern"
    >
        <global-events @keydown.tab.capture="forceCloseMoveAutocomplete" />
        <template slot="button-content">
            <bm-icon icon="folder" size="2x" />
            <span class="d-none d-lg-block"> {{ $t("mail.actions.move") }}</span>
        </template>
        <bm-dropdown-autocomplete
            ref="moveAutocomplete"
            v-slot="{ item }"
            v-model.trim="pattern"
            :items="matchingFolders"
            search-icon
            :max-results="maxFolders"
            has-divider-under-input
            @keydown.esc.native="resetPattern"
        >
            <bm-dropdown-item-button
                class="text-nowrap text-truncate w-100"
                :title="$tc('mail.actions.move.item', 1, { path: item.path })"
                @click="moveToFolder(item)"
            >
                <template #icon>
                    <mail-folder-icon no-text :shared="isFolderOfMailshare(item)" :folder="item" />
                </template>
                {{ translatePath(item.path) }}
            </bm-dropdown-item-button>
        </bm-dropdown-autocomplete>
        <bm-dropdown-divider />
        <bm-dropdown-form
            v-if="pattern === ''"
            class="new-folder position-relative"
            :aria-label="$t('mail.actions.create.folder')"
            :title="$t('mail.actions.create.folder')"
        >
            <mail-folder-input
                class="pl-2 pr-1"
                :submit-on-focusout="false"
                @submit="newFolderName => moveToFolder({ name: newFolderName, path: newFolderName })"
                @keydown.left.native.stop
                @keydown.right.native.stop
                @keydown.esc.native.stop
            />
        </bm-dropdown-form>
        <bm-dropdown-item-button
            v-else-if="displayCreateFolderBtnFromPattern"
            :aria-label="$tc('mail.actions.move.item', 1, { path: pattern })"
            :title="$tc('mail.actions.move.item', 1, { path: pattern })"
            icon="plus"
            @click="moveToFolder({ name: pattern, path: pattern })"
        >
            {{ $t("mail.folder.new.from_pattern", [pattern]) }}
        </bm-dropdown-item-button>
    </bm-dropdown>
</template>

<script>
import {
    BmDropdown,
    BmDropdownAutocomplete,
    BmDropdownDivider,
    BmDropdownForm,
    BmDropdownItemButton,
    BmIcon
} from "@bluemind/styleguide";
import { mapGetters, mapState } from "vuex";
import GlobalEvents from "vue-global-events";
import MailFolderIcon from "../../MailFolderIcon";
import MailFolderInput from "../../MailFolderInput";
import { MailboxType } from "~model/mailbox";
import { isNameValid, translatePath } from "~model/folder";
import { ACTIVE_MESSAGE, MY_INBOX, MY_TRASH, FOLDERS_BY_UPPERCASE_PATH } from "~getters";
import { MoveMixin } from "~mixins";

export default {
    name: "MailToolbarConsultMessageMoveAction",
    components: {
        BmDropdown,
        BmDropdownAutocomplete,
        BmDropdownDivider,
        BmDropdownForm,
        BmDropdownItemButton,
        BmIcon,
        GlobalEvents,
        MailFolderIcon,
        MailFolderInput
    },
    mixins: [MoveMixin],
    data() {
        return {
            maxFolders: 10,
            pattern: ""
        };
    },
    computed: {
        ...mapGetters("mail-webapp", ["nextMessageKey"]),
        ...mapState("mail", ["activeFolder", "folders", "mailboxes", "messages", "selection"]),
        ...mapGetters("mail", { ACTIVE_MESSAGE, MY_TRASH, MY_INBOX, FOLDERS_BY_UPPERCASE_PATH }),
        displayCreateFolderBtnFromPattern() {
            let pattern = this.pattern;
            if (pattern !== "") {
                pattern = pattern.replace(/\/+/, "/").replace(/^\/?(.*)\/?$/g, "$1");
                return pattern && isNameValid(pattern, pattern, this.FOLDERS_BY_UPPERCASE_PATH) === true;
            }
            return false;
        },
        matchingFolders() {
            if (this.pattern !== "") {
                const filtered = Object.values(this.folders).filter(
                    folder =>
                        folder.key !== this.activeFolder &&
                        folder.writable &&
                        (folder.path.toLowerCase().includes(this.pattern.toLowerCase()) ||
                            folder.name.toLowerCase().includes(this.pattern.toLowerCase()))
                );
                if (filtered) {
                    return filtered.slice(0, this.maxFolders);
                }
            }
            return [this.MY_INBOX, this.MY_TRASH].filter(folder => folder && folder.key !== this.activeFolder);
        }
    },
    methods: {
        moveToFolder(folder) {
            const selectedMessages = this.selection.map(key => this.messages[key]);
            const toBeMoved = selectedMessages.length > 0 ? selectedMessages : this.ACTIVE_MESSAGE;
            this.MOVE_MESSAGES({ messages: toBeMoved, folder });
            this.$refs["move-dropdown"].hide(true);
        },
        openMoveAutocomplete() {
            this.$nextTick(() => this.$refs["moveAutocomplete"].focus());
        },
        forceCloseMoveAutocomplete() {
            this.$refs["move-dropdown"].hide(true);
            this.pattern = "";
        },
        resetPattern() {
            this.pattern = "";
        },
        isFolderOfMailshare(folder) {
            return this.mailboxes[folder.mailboxRef.key].type === MailboxType.MAILSHARE;
        },
        translatePath(path) {
            return translatePath(path);
        },
        mailbox(folder) {
            return this.mailboxes[folder.mailboxRef.key];
        }
    }
};
</script>

<style lang="scss">
@import "~@bluemind/styleguide/css/variables";
@import "~@bluemind/styleguide/css/mixins";

.move-message {
    input {
        border: none !important;
    }

    .dropdown-menu {
        min-width: 20vw;
        max-width: 40vw;
    }

    .bm-dropdown-autocomplete form {
        input {
            padding-left: $sp-3;
        }
        padding: 0 !important;
    }

    .dropdown-divider {
        margin: 0 !important;
    }

    .dropdown-item-content {
        @include text-overflow;
    }

    .dropdown-item,
    .b-dropdown-form {
        &:hover {
            background-color: $extra-light !important;
        }

        &:focus,
        &:focus:hover {
            background-color: $surface-active-bg !important;
        }
    }

    .b-dropdown-form {
        padding-top: $sp-1;
    }

    .new-folder .b-dropdown-form {
        padding-right: 0;
        padding-left: 0;
        padding-bottom: 0;
    }
}
</style>
