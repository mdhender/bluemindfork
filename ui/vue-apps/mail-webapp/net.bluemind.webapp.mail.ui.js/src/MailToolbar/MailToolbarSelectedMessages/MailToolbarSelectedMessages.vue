<template>
    <div class="mail-toolbar-consult-message">
        <bm-button
            v-show="displayMarkAsRead"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="unread"
            :title="$tc('mail.actions.mark_read.aria', selectedMessageKeys.length || 1)"
            :aria-label="$tc('mail.actions.mark_read.aria', selectedMessageKeys.length || 1)"
            @click="doMarkAsRead"
        >
            <bm-icon icon="read" size="2x" />
            <span class="d-none d-lg-block"> {{ $tc("mail.actions.mark_read", selectedMessageKeys.length || 1) }}</span>
        </bm-button>
        <bm-button
            v-show="displayMarkAsUnread"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="read"
            :title="$tc('mail.actions.mark_unread.aria', selectedMessageKeys.length || 1)"
            :aria-label="$tc('mail.actions.mark_unread.aria', selectedMessageKeys.length || 1)"
            @click="doMarkAsUnread"
        >
            <bm-icon icon="unread" size="2x" />
            <span class="d-none d-lg-block">{{
                $tc("mail.actions.mark_unread", selectedMessageKeys.length || 1)
            }}</span>
        </bm-button>
        <mail-toolbar-selected-messages-move-action v-show="!isSelectionMultiple && !selectionHasReadOnlyFolders" />
        <bm-button
            v-show="!selectionHasReadOnlyFolders"
            v-bm-tooltip.bottom
            variant="simple-dark"
            :title="$tc('mail.actions.remove.aria')"
            :aria-label="$tc('mail.actions.remove.aria')"
            @click.exact="remove"
            @click.shift.exact="purge"
        >
            <bm-icon icon="trash" size="2x" />
            <span class="d-none d-lg-block">{{ $tc("mail.actions.remove") }}</span>
        </bm-button>
        <bm-button
            v-show="displayMarkAsFlagged"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="flagged"
            :title="$tc('mail.actions.mark_flagged.aria', selectedMessageKeys.length)"
            :aria-label="$tc('mail.actions.mark_flagged.aria', selectedMessageKeys.length)"
            @click="doMarkAsFlagged"
        >
            <bm-icon icon="flag-outline" size="2x" />
            <span class="d-none d-lg-block"> {{ $tc("mail.actions.mark_flagged") }}</span>
        </bm-button>
        <bm-button
            v-show="displayMarkAsUnflagged"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="unflagged"
            :title="$tc('mail.actions.mark_unflagged.aria', selectedMessageKeys.length)"
            :aria-label="$tc('mail.actions.mark_unflagged.aria', selectedMessageKeys.length)"
            @click="doMarkAsUnflagged"
        >
            <bm-icon icon="flag-fill" size="2x" class="text-warning" />
            <span class="d-none d-lg-block"> {{ $tc("mail.actions.mark_as_unflagged") }}</span>
        </bm-button>
        <mail-toolbar-selected-messages-other-actions v-if="!selectionHasReadOnlyFolders" />
    </div>
</template>

<script>
import { BmButton, BmIcon, BmTooltip } from "@bluemind/styleguide";
import { ItemUri } from "@bluemind/item-uri";
import { mapActions, mapGetters, mapState } from "vuex";
import MailToolbarSelectedMessagesMoveAction from "./MailToolbarSelectedMessagesMoveAction";
import MailToolbarSelectedMessagesOtherActions from "./MailToolbarSelectedMessagesOtherActions";
import { Flag } from "@bluemind/email";

export default {
    name: "MailToolbarSelectedMessages",
    components: {
        BmButton,
        BmIcon,
        MailToolbarSelectedMessagesMoveAction,
        MailToolbarSelectedMessagesOtherActions
    },
    directives: { BmTooltip },
    computed: {
        ...mapState("mail-webapp", ["currentFolderKey", "selectedMessageKeys"]),
        ...mapGetters("mail-webapp", [
            "areAllMessagesSelected",
            "areAllSelectedMessagesFlagged",
            "areAllSelectedMessagesRead",
            "areAllSelectedMessagesUnflagged",
            "areAllSelectedMessagesUnread",
            "areMessagesFiltered",
            "isSearchMode",
            "my",
            "nextMessageKey"
        ]),
        ...mapGetters("mail-webapp/currentMessage", { currentMessage: "message" }),
        ...mapState("mail", ["folders"]),
        isSelectionMultiple() {
            return this.selectedMessageKeys.length > 1;
        },
        displayMarkAsRead() {
            if (this.isSelectionMultiple) {
                return !this.areAllSelectedMessagesRead;
            } else {
                return this.currentMessage.states.includes("not-seen");
            }
        },
        displayMarkAsUnread() {
            if (this.isSelectionMultiple) {
                return !this.areAllSelectedMessagesUnread;
            } else {
                return !this.currentMessage.states.includes("not-seen");
            }
        },
        displayMarkAsFlagged() {
            if (this.selectionHasReadOnlyFolders) {
                return false;
            } else if (this.isSelectionMultiple) {
                return !this.areAllSelectedMessagesFlagged;
            } else {
                return !this.currentMessage.flags.includes(Flag.FLAGGED);
            }
        },
        displayMarkAsUnflagged() {
            if (this.selectionHasReadOnlyFolders) {
                return false;
            } else if (this.isSelectionMultiple) {
                return !this.areAllSelectedMessagesUnflagged;
            } else {
                return this.currentMessage.flags.includes(Flag.FLAGGED);
            }
        },
        selectionHasReadOnlyFolders() {
            const selection = this.selectedMessageKeys.length ? this.selectedMessageKeys : [this.currentMessage.key];
            return selection.some(messageKey => !this.folders[ItemUri.container(messageKey)].writable);
        }
    },
    methods: {
        ...mapActions("mail-webapp", {
            markAsUnread: "markAsUnread",
            markFolderAsRead: "markFolderAsRead",
            markMessagesAsFlagged: "markAsFlagged",
            markMessagesAsRead: "markAsRead",
            markMessagesAsUnflagged: "markAsUnflagged"
        }),
        async purge() {
            const confirm = await this.$bvModal.msgBoxConfirm(
                this.$tc("mail.actions.purge.modal.content", this.selectedMessageKeys.length || 1, {
                    subject: this.currentMessage && this.currentMessage.subject
                }),
                {
                    title: this.$tc("mail.actions.purge.modal.title", this.selectedMessageKeys.length || 1),
                    okTitle: this.$t("common.delete"),
                    cancelVariant: "outline-secondary",
                    cancelTitle: this.$t("common.cancel"),
                    centered: true,
                    hideHeaderClose: false
                }
            );
            if (confirm) {
                // do this before followed async operations
                const nextMessageKey = this.nextMessageKey;
                this.$store.dispatch(
                    "mail-webapp/purge",
                    this.selectedMessageKeys.length > 1 ? this.selectedMessageKeys : this.currentMessage.key
                );
                if (!this.isSelectionMultiple) {
                    this.$router.navigate({ name: "v:mail:message", params: { message: nextMessageKey } });
                }
            }
        },
        remove() {
            if (this.currentFolderKey === this.my.TRASH.key) {
                this.purge();
            } else {
                // do this before followed async operations
                const nextMessageKey = this.nextMessageKey;
                this.$store.dispatch(
                    "mail-webapp/remove",
                    this.selectedMessageKeys.length > 1 ? this.selectedMessageKeys : this.currentMessage.key
                );
                if (!this.isSelectionMultiple) {
                    this.$router.navigate({ name: "v:mail:message", params: { message: nextMessageKey } });
                }
            }
        },
        selectedKeys() {
            return this.isSelectionMultiple ? this.selectedMessageKeys : [this.currentMessage.key];
        },
        doMarkAsRead() {
            const areAllMessagesInFolderSelected =
                this.areAllMessagesSelected && !this.areMessagesFiltered && !this.isSearchMode;
            areAllMessagesInFolderSelected
                ? this.markFolderAsRead(this.currentFolderKey)
                : this.markMessagesAsRead(this.selectedKeys());
        },
        doMarkAsUnread() {
            if (this.isSelectionMultiple) {
                this.markAsUnread(this.selectedMessageKeys);
            } else {
                this.markAsUnread([this.currentMessage.key]);
            }
        },
        doMarkAsFlagged() {
            this.markMessagesAsFlagged(this.selectedKeys());
        },
        doMarkAsUnflagged() {
            this.markMessagesAsUnflagged(this.selectedKeys());
        }
    }
};
</script>

<style lang="scss">
@import "~@bluemind/styleguide/css/_variables";

.mail-toolbar-consult-message .unread,
.mail-toolbar-consult-message .read,
.mail-toolbar-consult-message .flagged,
.mail-toolbar-consult-message .unflagged {
    @media (min-width: map-get($grid-breakpoints, "lg")) {
        width: 8rem;
    }
}
</style>
