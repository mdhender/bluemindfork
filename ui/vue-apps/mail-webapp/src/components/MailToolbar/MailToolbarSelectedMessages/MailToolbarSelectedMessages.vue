<template>
    <div class="mail-toolbar-consult-message">
        <bm-button
            v-show="displayMarkAsRead"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="unread"
            :title="$tc('mail.actions.mark_read.aria', selection.length || 1)"
            :aria-label="$tc('mail.actions.mark_read.aria', selection.length || 1)"
            @click="doMarkAsRead"
        >
            <bm-icon icon="read" size="2x" />
            <span class="d-none d-lg-block"> {{ $tc("mail.actions.mark_read", selection.length || 1) }}</span>
        </bm-button>
        <bm-button
            v-show="displayMarkAsUnread"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="read"
            :title="$tc('mail.actions.mark_unread.aria', selection.length || 1)"
            :aria-label="$tc('mail.actions.mark_unread.aria', selection.length || 1)"
            @click="doMarkAsUnread"
        >
            <bm-icon icon="unread" size="2x" />
            <span class="d-none d-lg-block">{{ $tc("mail.actions.mark_unread", selection.length || 1) }}</span>
        </bm-button>
        <mail-toolbar-selected-messages-move-action
            v-show="!MULTIPLE_MESSAGE_SELECTED && !selectionHasReadOnlyFolders"
        />
        <bm-button
            v-show="!selectionHasReadOnlyFolders"
            v-bm-tooltip.bottom
            variant="simple-dark"
            :title="$tc('mail.actions.remove.aria')"
            :aria-label="$tc('mail.actions.remove.aria')"
            @click.exact="remove"
            @click.shift.exact="purge"
        >
            <bm-icon icon="trash" size="2x" />
            <span class="d-none d-lg-block">{{ $tc("mail.actions.remove") }}</span>
        </bm-button>
        <bm-button
            v-show="displayMarkAsFlagged"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="flagged"
            :title="$tc('mail.actions.mark_flagged.aria', selection.length)"
            :aria-label="$tc('mail.actions.mark_flagged.aria', selection.length)"
            @click="doMarkAsFlagged"
        >
            <bm-icon icon="flag-outline" size="2x" />
            <span class="d-none d-lg-block"> {{ $tc("mail.actions.mark_flagged") }}</span>
        </bm-button>
        <bm-button
            v-show="displayMarkAsUnflagged"
            v-bm-tooltip.bottom
            variant="simple-dark"
            class="unflagged"
            :title="$tc('mail.actions.mark_unflagged.aria', selection.length)"
            :aria-label="$tc('mail.actions.mark_unflagged.aria', selection.length)"
            @click="doMarkAsUnflagged"
        >
            <bm-icon icon="flag-fill" size="2x" class="text-warning" />
            <span class="d-none d-lg-block"> {{ $tc("mail.actions.mark_as_unflagged") }}</span>
        </bm-button>
        <mail-toolbar-selected-messages-other-actions v-if="!selectionHasReadOnlyFolders" />
    </div>
</template>

<script>
import { mapActions, mapGetters, mapState } from "vuex";
import { BmButton, BmIcon, BmTooltip } from "@bluemind/styleguide";
import { ItemUri } from "@bluemind/item-uri";
import { Flag } from "@bluemind/email";
import MailToolbarSelectedMessagesMoveAction from "./MailToolbarSelectedMessagesMoveAction";
import MailToolbarSelectedMessagesOtherActions from "./MailToolbarSelectedMessagesOtherActions";
import {
    ALL_MESSAGES_ARE_SELECTED,
    ALL_SELECTED_MESSAGES_ARE_FLAGGED,
    ALL_SELECTED_MESSAGES_ARE_READ,
    ALL_SELECTED_MESSAGES_ARE_UNFLAGGED,
    ALL_SELECTED_MESSAGES_ARE_UNREAD,
    MESSAGE_LIST_FILTERED,
    MESSAGE_LIST_IS_SEARCH_MODE,
    MULTIPLE_MESSAGE_SELECTED,
    MY_TRASH
} from "~getters";

export default {
    name: "MailToolbarSelectedMessages",
    components: {
        BmButton,
        BmIcon,
        MailToolbarSelectedMessagesMoveAction,
        MailToolbarSelectedMessagesOtherActions
    },
    directives: { BmTooltip },
    computed: {
        ...mapGetters("mail-webapp", ["nextMessageKey"]),
        ...mapGetters("mail", {
            ALL_MESSAGES_ARE_SELECTED,
            ALL_SELECTED_MESSAGES_ARE_FLAGGED,
            ALL_SELECTED_MESSAGES_ARE_READ,
            ALL_SELECTED_MESSAGES_ARE_UNFLAGGED,
            ALL_SELECTED_MESSAGES_ARE_UNREAD,
            MULTIPLE_MESSAGE_SELECTED,
            MY_TRASH,
            MESSAGE_LIST_FILTERED,
            MESSAGE_LIST_IS_SEARCH_MODE
        }),
        ...mapState("mail-webapp/currentMessage", { currentMessageKey: "key" }),
        ...mapState("mail", ["folders", "activeFolder", "messages", "selection"]),
        message() {
            return this.messages[this.currentMessageKey];
        },
        displayMarkAsRead() {
            if (this.MULTIPLE_MESSAGE_SELECTED) {
                return !this.ALL_SELECTED_MESSAGES_ARE_READ;
            } else {
                return !this.message.flags.includes(Flag.SEEN);
            }
        },
        displayMarkAsUnread() {
            if (this.MULTIPLE_MESSAGE_SELECTED) {
                return !this.ALL_SELECTED_MESSAGES_ARE_UNREAD;
            } else {
                return this.message.flags.includes(Flag.SEEN);
            }
        },
        displayMarkAsFlagged() {
            if (this.selectionHasReadOnlyFolders) {
                return false;
            } else if (this.MULTIPLE_MESSAGE_SELECTED) {
                return !this.ALL_SELECTED_MESSAGES_ARE_FLAGGED;
            } else {
                return !this.message.flags.includes(Flag.FLAGGED);
            }
        },
        displayMarkAsUnflagged() {
            if (this.selectionHasReadOnlyFolders) {
                return false;
            } else if (this.MULTIPLE_MESSAGE_SELECTED) {
                return !this.ALL_SELECTED_MESSAGES_ARE_UNFLAGGED;
            } else {
                return this.message.flags.includes(Flag.FLAGGED);
            }
        },
        selectionHasReadOnlyFolders() {
            const selection = this.selection.length ? this.selection : [this.currentMessageKey];
            return selection.some(messageKey => !this.folders[ItemUri.container(messageKey)].writable);
        }
    },
    methods: {
        ...mapActions("mail-webapp", {
            markAsUnread: "markAsUnread",
            markFolderAsRead: "markFolderAsRead",
            markMessagesAsFlagged: "markAsFlagged",
            markMessagesAsRead: "markAsRead",
            markMessagesAsUnflagged: "markAsUnflagged"
        }),
        async purge() {
            const confirm = await this.$bvModal.msgBoxConfirm(
                this.$tc("mail.actions.purge.modal.content", this.selection.length || 1, {
                    subject: this.message && this.message.subject
                }),
                {
                    title: this.$tc("mail.actions.purge.modal.title", this.selection.length || 1),
                    okTitle: this.$t("common.delete"),
                    cancelVariant: "outline-secondary",
                    cancelTitle: this.$t("common.cancel"),
                    centered: true,
                    hideHeaderClose: false
                }
            );
            if (confirm) {
                // do this before followed async operations
                const nextMessageKey = this.nextMessageKey;
                this.$store.dispatch(
                    "mail-webapp/purge",
                    this.selection.length > 1 ? this.selection : this.currentMessageKey
                );
                if (!this.MULTIPLE_MESSAGE_SELECTED) {
                    this.$router.navigate({ name: "v:mail:message", params: { message: nextMessageKey } });
                }
            }
        },
        remove() {
            if (this.activeFolder === this.MY_TRASH.key) {
                this.purge();
            } else {
                // do this before followed async operations
                const nextMessageKey = this.nextMessageKey;
                this.$store.dispatch(
                    "mail-webapp/remove",
                    this.selection.length > 1 ? this.selection : this.currentMessageKey
                );
                if (!this.MULTIPLE_MESSAGE_SELECTED) {
                    this.$router.navigate({ name: "v:mail:message", params: { message: nextMessageKey } });
                }
            }
        },
        selectedKeys() {
            return this.MULTIPLE_MESSAGE_SELECTED ? this.selection : [this.currentMessageKey];
        },
        doMarkAsRead() {
            const areAllMessagesInFolderSelected =
                this.ALL_MESSAGES_ARE_SELECTED && !this.MESSAGE_LIST_FILTERED && !this.MESSAGE_LIST_IS_SEARCH_MODE;
            areAllMessagesInFolderSelected
                ? this.markFolderAsRead(this.activeFolder)
                : this.markMessagesAsRead(this.selectedKeys());
        },
        doMarkAsUnread() {
            if (this.MULTIPLE_MESSAGE_SELECTED) {
                this.markAsUnread(this.selection);
            } else {
                this.markAsUnread([this.currentMessageKey]);
            }
        },
        doMarkAsFlagged() {
            this.markMessagesAsFlagged(this.selectedKeys());
        },
        doMarkAsUnflagged() {
            this.markMessagesAsUnflagged(this.selectedKeys());
        }
    }
};
</script>
