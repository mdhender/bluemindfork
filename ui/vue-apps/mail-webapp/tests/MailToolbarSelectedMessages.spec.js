import MailToolbarSelectedMessages from "../src/components/MailToolbar/MailToolbarSelectedMessages";
jest.mock("@bluemind/styleguide/css/_variables.scss", () => ({ iconsColors: "" }));
import { createWrapper } from "./testUtils";

describe("MailToolbarSelectedMessages", () => {
    test("is a Vue instance", () => {
        const wrapper = createWrapper(MailToolbarSelectedMessages);
        expect(wrapper.vm).toBeTruthy();
    });

    test("should match snapshot", () => {
        const wrapper = createWrapper(MailToolbarSelectedMessages);
        expect(wrapper.element).toMatchSnapshot();
    });
});
