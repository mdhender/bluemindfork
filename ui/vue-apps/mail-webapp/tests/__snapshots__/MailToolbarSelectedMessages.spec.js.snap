// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailToolbarSelectedMessages should match snapshot 1`] = `
<div
  class="mail-toolbar-consult-message"
>
  <!---->
</div>
`;
